libsolidity/ast/Types.cpp

```diff
@@ -1992,16 +2011,47 @@ bool StructType::operator==(Type const& _other) const
 	return ReferenceType::operator==(other) && other.m_struct == m_struct;
 }
 
+
 unsigned StructType::calldataEncodedSize(bool) const
 {
+	solAssert(!isDynamicallyEncoded(), "");
+
+	unsigned size = 0;
+	for (auto const& member: members(nullptr))
+		if (!member.type->canLiveOutsideStorage())
+			return 0;
+		else
+		{
+			// Struct members are always padded.
+			unsigned memberSize = member.type->calldataEncodedSize();
+			if (memberSize == 0)
+				return 0;
+			size += memberSize;
+		}
+	return size;
+}
+
+
+unsigned StructType::calldataHeadIncrement() const
+{
+	if (isDynamicallyEncoded())
+		return 32;
+	else
+		return calldataEncodedSize(true);
+}
+
+unsigned StructType::calldataEncodedTailSize() const
+{
+	solAssert(isDynamicallyEncoded(), "");
+
 	unsigned size = 0;
 	for (auto const& member: members(nullptr))
 		if (!member.type->canLiveOutsideStorage())
 			return 0;
 		else
 		{
 			// Struct members are always padded.
-			unsigned memberSize = member.type->calldataEncodedSize(true);
+			unsigned memberSize = member.type->calldataHeadIncrement();
 			if (memberSize == 0)
```

Comment:
Could you add a comment to explain why it `return 0;` when member size is 0? (I don't get it)

Reply:
It might be better to assert that ``memberSize != 0`` here - I think that should work.

Currently ``0`` was returned to indicate that the tail cannot be encoded in calldata, which is the case if any member cannot be encoded in calldata (i.e. when any member has size 0).

Reply:
Now the case vanished, since I changed the functions to throw whenever a type cannot be encoded in calldata.